src/sparse_blas/enum_data_types.hpp

```diff
@@ -0,0 +1,69 @@
+/***************************************************************************
+*  Copyright (C) Codeplay Software Limited
+*  Licensed under the Apache License, Version 2.0 (the "License");
+*  you may not use this file except in compliance with the License.
+*  You may obtain a copy of the License at
+*
+*      http://www.apache.org/licenses/LICENSE-2.0
+*
+*  For your convenience, a copy of the License has been included in this
+*  repository.
+*
+*  Unless required by applicable law or agreed to in writing, software
+*  distributed under the License is distributed on an "AS IS" BASIS,
+*  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+*  See the License for the specific language governing permissions and
+*  limitations under the License.
+*
+**************************************************************************/
+
+#ifndef _ONEMKL_SRC_SPARSE_BLAS_ENUM_DATA_TYPES_HPP_
+#define _ONEMKL_SRC_SPARSE_BLAS_ENUM_DATA_TYPES_HPP_
+
+#include <string>
+
+namespace oneapi::mkl::sparse::detail {
+
+enum data_type { none, int32, int64, real_fp32, real_fp64, complex_fp32, complex_fp64 };
+
+inline std::string data_type_to_str(data_type data_type) {
+    switch (data_type) {
+        case none: return "none";
+        case int32: return "int32";
+        case int64: return "int64";
+        case real_fp32: return "real_fp32";
+        case real_fp64: return "real_fp64";
+        case complex_fp32: return "complex_fp32";
+        case complex_fp64: return "complex_fp64";
+        default: return "unknown";
+    }
+}
+
+template <typename T>
+data_type get_data_type() {
```

Comment:
Don't you need
```suggestion
constexpr data_type get_data_type() {
```
to ensure the function is `constexpr` and not just its internals?

Reply:
Technically it is not needed because `get_data_type` is not used in a context were it is evaluated at compile time. The internals didn't really need the `constexpr` either, this is just a force of habit whenever I use type traits.
It doesn't hurt to make it `constexpr` so done in https://github.com/oneapi-src/oneMKL/pull/500/commits/d04452ac2c020f5c0336d05290d402afeb123713

Reply:
Resoving old comment